fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/plans/FDBComparatorPlanTest.java

```diff
@@ -0,0 +1,314 @@
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 1);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(2)),
+                context -> assertDiscardedAtMost(134, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoDifferentInnerPlansOneEndsSooner() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        // This plan has same records as the previous one, but end sooner
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.and(
+                        Query.field("num_value_2").equalsValue(1),
+                        Query.field("rec_no").lessThan(5L)))
+                .build();
+
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 0);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(134, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoDifferentInnerPlansReferenceEndsSooner() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        // This plan has same records as the previous one, but end sooner
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.and(
+                        Query.field("num_value_2").equalsValue(1),
+                        Query.field("rec_no").lessThan(5L)))
+                .build();
+
+        // Note this time the reference plan is hte second one
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 1);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(10, context));
+        assertEquals(2, count);
+    }
+
+    @DualPlannerTest
+    void testNoInnerPlansFail() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        assertThrows(VerifyException.class, () -> RecordQueryComparatorPlan.from(Collections.emptyList(), primaryKey(), 0));
+    }
+
+    @DualPlannerTest
+    void testTwoSameInnerPlansWithContinuation() throws Throwable {
```

Comment:
It would probably be good to have some versions of this that failed with out-of-band limits. I think the rows scanned limit should be deterministic enough that if it were used, it wouldn't cause the resulting tests to be flaky

Reply:
Done.